lib/mayaUsd/ufe/UsdHierarchy.cpp

```diff
@@ -117,14 +120,15 @@ Ufe::SceneItemList UsdHierarchy::createUFEChildList(const UsdPrimSiblingRange& r
     // Return UFE child list from input USD child list.
     Ufe::SceneItemList children;
     for (const auto& child : range) {
-        children.emplace_back(UsdSceneItem::create(fItem->path() + child.GetName(), child));
```

Comment:
Not sure I understand why we need the stripInstanceIndexFromUfePath() call.  Is it the case that a point instancer instance can have children?  Because here we're in the parent, collecting children from USD.

Reply:
Hmm, I may be misunderstanding when some of this `UsdHierarchy` API gets invoked then. I was assuming that we might get into situations where this instance of `UsdHierarchy` has an `fItem` that represents a point instance, but maybe that's not actually possible?

When a `UsdSceneItem` represents a point instance, its `prim()` member will return the `PointInstancer` prim that created the instance. There is no representation of the instance itself anywhere in USD. As a result, I'm not sure that it makes sense to think of a point instance as having a parent. Instead, there's a `PointInstancer` prim that created it.

So my goal then in adding these `stripInstanceIndexFromUfePath()` calls was to ensure that if we somehow are in a case where this `UsdHierarchy`'s `fItem` is a point instance, we **always** work in terms of the `PointInstancer` prim instead when computing parents or children, since the point instance really has neither and is more a product of imaging.

Maybe these overrides should be detecting when `fItem` is a point instance and return `nullptr` or  empty lists or the equivalent instead?

Reply:
O.K., let me see if I have my head wrapped around the UFE data model for point instancers and point instances --- and I still have a question :)
- A point instance UFE scene item has no parent, and has no children.  Not having a parent is a departure from existing UFE practice, which may uncover latent bugs, but retrieving a parent should go through the Ufe::Hierarchy interface, which you'll be defining.  Returning a nullptr parent and an empty child list is perfect.
- The prim of a point instance UFE scene item is its point instancer prim.  This makes sense.
- What are the children of a point instancer?
We should make sure all of this is as obvious as possible in the implementation and comments.
-

Reply:
Yeah, all good insights and questions. Hopefully my understanding is starting to congeal here too. :)

On the USD side, when we're talking about parents or children, we are virtually always talking about **namespace** parents or **namespace** children. These would be defined prims with unique locations in USD namespace. Point instances have no such location in USD namespace. They are instead generated based on data members of a `PointInstancer` prim and a "prototype" that is being instanced, where the prototype refers to some location in USD namespace which may or may not be a hierarchy of prims. In terms of USD, a point instance itself doesn't really have either a parent or children then. A `PointInstancer` is an actual prim with namespace parents, so it clearly has a parent, but the point instances that it generates would **not** be considered its children. It's common for `PointInstancer` prims to have a `prototypes` scope as a child where all of its prototypes are defined, so it's really only proper USD prims that would be considered children of a `PointInstancer`.

On the UFE side, I think carrying over the child-less property of point instances from USD makes sense, so there is no way to traverse down from a point instance. It **is** useful to be able to get from a point instance to the `PointInstancer` that generated it though, in particular for pick-walking, so I think it does make sense to have UFE report that the "parent" of a point instance is its `PointInstancer`. For the `PointInstancer` itself, their hierarchy should be treated the same way as any other non-`PointInstancer` prim, where only USD namespace parents and children are considered UFE parents and children.

I re-worked this a bit and tried to make this a bit clearer with some additional comments.

Reply:
I took another crack at this in 95a4d4174531e488facd72c1789ebef3434ff469. Let me know what you think!

I also added a pick-walking test in 9c5099586fc182b547eb306d728329634e2de3bc.

Reply:
Sounds good. I just added some documentation to UsdSceneItem.h to capture everything we've been discussing here. You should see that in a71c3dd30a13c12c96ebe076bca90779f299fe37. Give it a read through and let me know what you think.

I added that in a rebase (apologies, I know that's been somewhat complicating reviewing this), so although the commits were rewritten, nothing else was changed.

Reply:
Love it, very thorough, thanks.  Now the only challenge is to get people to RTFM, but we have something to point them to.